server/src/main/scala/org/scalaexercises/evaluator/evaluation.scala

```diff
@@ -96,68 +108,60 @@ class Evaluator[F[_]: Sync](timeout: FiniteDuration = 20.seconds)(
         }
       }
 
-      override lazy val compilerMessageHandler: Option[Reporter] = Some(new AbstractReporter {
-        override val settings: Settings    = compilerSettings
-        override def displayPrompt(): Unit = ()
-        override def display(pos: Position, msg: String, severity: this.type#Severity): Unit =
+      override lazy val compilerMessageHandler: Option[Reporter] = Some(new FilteringReporter {
+        override def settings: Settings = compilerSettings
+
+        override def doReport(pos: Position, msg: String, severity: Severity): Unit =
           errors += convert((pos, msg, severity.toString))
+
         override def reset() = {
           super.reset()
           errors = Map.empty
         }
       })
+
     }
-  }
 
-  private[this] def evaluate[T](code: String, jars: Seq[File]): EvalResult[T] = {
-    val eval = createEval(jars)
+  private[this] def evaluate[T](code: String, jars: Seq[File]): F[EvalResult[T]] = {
+    F.bracket(F.delay(createEval(jars))) { evalInstance =>
+      val outCapture = new ByteArrayOutputStream
 
-    val outCapture = new ByteArrayOutputStream
+      F.delay[EvalResult[T]](Console.withOut(outCapture) {
 
-    Console.withOut(outCapture) {
+        val result = Try(evalInstance.execute[T](code, resetState = true, jars = jars))
 
-      val result = for {
-        _      ← Try(eval.check(code))
-        result ← Try(eval.execute[T](code, resetState = true, jars = jars))
-      } yield result
+        val errors = evalInstance.errors
 
-      val errors = eval.errors
+        result match {
+          case scala.util.Success(r) => EvalSuccess[T](errors, r, outCapture.toString)
+          case scala.util.Failure(t) =>
+            t match {
+              case e: CompilerException => CompilationError(errors)
+              case NonFatal(e) =>
+                EvalRuntimeError(errors, Option(RuntimeError(e, None)))
+              case e => GeneralError(e)
+            }
+        }
+      })
+    }(EI => F.delay(EI.clean()))
 
-      result match {
-        case scala.util.Success(r) ⇒ EvalSuccess[T](errors, r, outCapture.toString)
-        case scala.util.Failure(t) ⇒
-          t match {
-            case e: CompilerException ⇒ CompilationError(errors)
-            case NonFatal(e) ⇒
-              EvalRuntimeError(errors, Option(RuntimeError(e, None)))
-            case e ⇒ GeneralError(e)
-          }
-      }
-    }
   }
 
   def eval[T](
       code: String,
       remotes: Seq[Remote] = Nil,
-      dependencies: Seq[Dependency] = Nil
+      dependencies: Seq[EvaluatorDependency] = Nil
   ): F[EvalResult[T]] = {
     for {
       allJars <- fetchArtifacts(remotes, dependencies)
       result <- allJars match {
         case Right(jars) =>
-          timeoutTo[EvalResult[T]](F.delay { evaluate(code, jars) }, timeout, Timeout[T](timeout))
+          evaluate(code, jars)
+            .timeoutTo(timeout, Timeout[T](timeout).asInstanceOf[EvalResult[T]].pure[F])
```

Comment:
I'm not pretty sure what you want to do here, shouldn't be something like this?

```scala
result <- allJars match {
        case Right(jars) =>
          evaluate[T](code, jars).timeout(timeout)
        case Left(fileError) => F.pure(UnresolvedDependency[T](fileError.describe))
      }
```

Reply:
When the execution reaches a timeout we want to return an `EvalResult`. The `timeout` method throws an exception when the execution takes more than the specified time. Should I use `timeout` with `handleError`?